### PR TITLE
RFC: refactor the solver-facing API

### PR DESCRIPTION
**Pull request by odow**
This PR refactors the solver-facing API of LQOI.

I have chosen long, explict names in snake_case.

### Remaining decisions
- [x] the name of `lqs_char`
- [x] `get_rhs` or `get_right_hand_side``

- [x] whether to tag this as a separate package, or incorporate into MOI.
Pros of incorporating:  it is easier to incorporate changes (i.e. the upcoming redefinition of `ScalarAffineFunction` etc.
Cons of incorporating:  testing.

### CODE REVIEW

File: src/solver_interface.jl [hunk 1]
```diff
@@ -78,23 +78,23 @@ Initializes a model given a model type `M` and an `env` that might be a `nothing
...
+#
+# """
+#     lqs_setlogfile!(m, path::String)::Void
+#
+# Set the log file to `path` for the model `m`.
+# """
+# function lqs_setlogfile!(m::LinQuadOptimizer, path) end
+#
+# # TODO(@joaquimg): what is this?
+# function lqs_getprobtype(m::LinQuadOptimizer) end
```
Comment (joaquimg) [inline src/solver_interface.jl]: you can remove these 3

**Comment by odow** [comment]
Okay @joaquimg, this is in a nicer state. Any objections to the renaming?

**Comment by joaquimg** [comment]
I will take a look today

File: src/solver_interface.jl [hunk 2]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
-function lqs_getrows(m::LinQuadOptimizer, row) end
-
-"""
-    lqs_getcoef(m, row::Int, col::Int)::Float64
+function get_linear_constraint end
+@deprecate lqs_getrows get_linear_constraint
 
-Get the linear coefficient of the variable in column `col`, constraint `row`.
 """
-function lqs_getcoef(m::LinQuadOptimizer, row, col) end
```
Comment (odow) [inline src/solver_interface.jl]: I got rid of this because I couldn't find it used anywhere

File: src/solver_interface.jl [hunk 3]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 
 A range constraint `l <= a'x <= u` is added as the linear constraint
-`a'x :RANGED l`, then this function is called to set `u - l`, the range value.
+`a'x [sym] l`, where `sym= lqs_char(m, Interval(l,u))`, then this function is called to set `u - l`, the range value.
 
-See `lqs_addrows!` for more.
+See `add_linear_constraints!` for more.
 """
-function lqs_chgrngval!(m::LinQuadOptimizer, rows, vals) end# later
+function change_range_value! end
```
Comment (odow) [inline src/solver_interface.jl]: I'm still open to an `add_ranged_constraint!` method instead of this.

File: src/solver_interface.jl [hunk 4]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 `:min` or `:max`.
 """
-function lqs_chgobjsen!(m::LinQuadOptimizer, sense) end
+function change_objective_sense! end
+@deprecate lqs_chgobjsen! change_objective_sense!
 
-#TODO(@joaquimg): why is this not in-place?
 """
-    lqs_getobj(m)::Vector{Float64}
+    get_linear_objective!(m, x::Vector{Float64})
```
Comment (odow) [inline src/solver_interface.jl]: I made this in-place.

File: src/solver_interface.jl [hunk 5]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 get the constraint primal `a'x + x'Qx` for each constraint, and store in `x`.
 `x` must have one element for each quadratic constraint.
 """
-function lqs_getqcax!(m::LinQuadOptimizer, x) end
+function get_quadratic_primal_solution! end
+@deprecate lqs_getqcax! get_quadratic_primal_solution!
 
 """
-    lqs_getdj!(m, x::Vector{Float64})
+    get_reducedcosts!(m, x::Vector{Float64})
```
Comment (odow) [inline src/solver_interface.jl]: Maybe this shoud be `get_variable_dual_solution!`

File: src/solver_interface.jl [hunk 6]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 Get the number of branch-and-cut nodes expolored during the most recent
 optimization of the model `m`.
 """
-function lqs_getnodecnt(m::LinQuadOptimizer) end
+function get_node_count end
+@deprecate lqs_getnodecnt get_node_count
 
 """
-    lqs_dualfarkas!(m, x::Vector{Float64})
+    get_farkasdual!(m, x::Vector{Float64})
```
Comment (odow) [inline src/solver_interface.jl]: For consistency `get_farkas_dual!`

File: src/solver_interface.jl [hunk 7]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 in the model `m`, and store in `x`. `x`must have one element for each linear
 constraint.
 """
-function lqs_dualfarkas!(m::LinQuadOptimizer, x) end
+function get_farkasdual! end
+@deprecate lqs_dualfarkas! get_farkasdual!
 
 """
-    lqs_getray!(m, x::Vector{Float64})
+    get_unboundedray!(m, x::Vector{Float64})
```
Comment (odow) [inline src/solver_interface.jl]: For consistency `get_unbounded_ray!`

File: src/solver_interface.jl [hunk 8]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 Get the unbounded ray (certificate of dual infeasiblility) for the linear constraints
 in the model `m`, and store in `x`. `x`must have one element for each variable.
 """
-function lqs_getray!(m::LinQuadOptimizer, x) end
+function get_unboundedray! end
+@deprecate lqs_getray! get_unboundedray!
 
 """
-    lqs_terminationstatus(m)
+    get_terminationstatus(m)
```
Comment (odow) [inline src/solver_interface.jl]: `get_termination_status`

File: src/solver_interface.jl [hunk 9]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 
 Get the termination status of the model `m`.
 """
-function lqs_terminationstatus(m::LinQuadOptimizer) end
+function get_terminationstatus end
+@deprecate lqs_terminationstatus get_terminationstatus
 
 """
-    lqs_primalstatus(m)
+    get_primalstatus(m)
```
Comment (odow) [inline src/solver_interface.jl]: `get_primal_status`

File: src/solver_interface.jl [hunk 10]
```diff
@@ -72,436 +72,469 @@ lqs_char(m::LinQuadOptimizer, ::Val{:Upperbound}) = Cchar('U')
...
 
 Get the primal status of the model `m`.
 """
-function lqs_primalstatus(m::LinQuadOptimizer) end
+function get_primalstatus end
+@deprecate lqs_primalstatus get_primalstatus
 
 """
-    lqs_dualstatus(m)
+    get_dualstatus(m)
```
Comment (odow) [inline src/solver_interface.jl]: `get_dual_status`

**Comment by joaquimg** [comment]
I will check getcoef usage in xpress. I agree with add_ranged.

**Comment by odow** [comment]
> I will check getcoef usage in xpress.

It doesn't matter if it is used in Xpress. It only matters if it is used in LQOI.

**Comment by joaquimg** [comment]
Yep, agree. But there is one function in LQOI that is extremely inefficient the one to query constraints and coefficients. This function might be a way to go that was forgotten in time and space.
Sure you can delete it and we add it back again if necessary.

**Comment by odow** [comment]
> Sure you can delete it and we add it back again if necessary.

This is one argument for keepling LQOI and MOI separate. MOI won't be changing but we might be changing LQOI...

**Comment by joaquimg** [comment]
Yes, there is performance tweaks to happen in LQOI, I think we should tag it separately (so its usable) and then we merge it into MOI

**Comment by joaquimg** [comment]
I am in favour of rhs, it is a bit different but I think its a well established jargon

**Comment by odow** [comment]
Okay, I've finished the renaming. Gurobi and GLPK both pass.

I propose merging this, registering LQOI in METADATA, then merging and tagging releases of both Gurobi and GLPK.

**Comment by mlubin** [comment]
@odow do you mean tagging Gurobi and GLPK with the LQOI dependency?

**Comment by odow** [comment]
> do you mean tagging Gurobi and GLPK with the LQOI dependency?

Yes.

I'm hestitant to merge LQOI into MOI in the short term as it is hard to test, and may yet change. Maybe we keep as separate until JuMP 0.19 comes out, then it should be stable enough to merge into MOI.

**Comment by mlubin** [comment]
It will be a bit annoying if we have to keep bumping all the MOI upper bounds if there are LQOI breaking changes. If you expect to do this more than twice per MOI bump then it's probably better to leave it as a separate package for now.

**Comment by joaquimg** [comment]
I think they might be breaking

**Comment by odow** [comment]
Let's punt on the decision for a bit. I'm going to merge this now, and then have a look at what needs to be implemented ASAP to avoid a lot of breaking changes.

**Comment by blegat** [comment]
I agree with @joaquimg and @odow , we need to at least wait until LQOI (same for SDOI) is able to test it by itself (e.g. without relying on an external solver). This might happen after JuMP v0.19, it would be nice if we don't delay JuMP v0.19 too much :)

**Comment by joaquimg** [comment]
@odow Do you plan anything else before merging?

**Pull request merged**